src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -66,169 +67,226 @@ public static string GetNuGetLockFilePath(string baseDirectory, string projectNa
-            var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
-            var lockFileRuntimes = nuGetLockFile.Targets.Select(t => t.RuntimeIdentifier).Where(r => r != null).Distinct();
+                    IEnumerable<LockFileDependency> directDependencies = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.Direct);
 
-            if (!projectRuntimesKeys.OrderedEquals(
-                            lockFileRuntimes,
-                            x => x,
-                            StringComparer.InvariantCultureIgnoreCase,
-                            StringComparer.InvariantCultureIgnoreCase))
-            {
-                return false;
-            }
+                    (var hasProjectDependencyChanged, var pmessage) = HasProjectDependencyChanged(framework.Dependencies, directDependencies, target.TargetFramework);
+                    if (hasProjectDependencyChanged)
+                    {
+                        // lock file is out of sync
+                        invalidReasons.Add(pmessage);
+                    }
 
-            // Validate all P2P references
-            foreach (var framework in project.RestoreMetadata.TargetFrameworks)
-            {
-                var target = nuGetLockFile.Targets.FirstOrDefault(
-                    t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
+                    var transitiveDependenciesEnforcedByCentralVersions = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.CentralTransitive).ToList();
+                    var transitiveDependencies = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.Transitive).ToList();
 
-                if (target == null)
-                {
-                    // a new target found in the dgSpec so invalidate existing lock file.
-                    return false;
+                    (var hasTransitiveDependencyChanged, var tmessage) = HasProjectTransitiveDependencyChanged(framework.CentralPackageVersions, transitiveDependenciesEnforcedByCentralVersions, transitiveDependencies);
+                    if (hasTransitiveDependencyChanged)
+                    {
+                        // lock file is out of sync
+                        invalidReasons.Add(tmessage);
+                    }
                 }
 
-                var queue = new Queue<Tuple<string, string>>();
-                var visitedP2PReference = new HashSet<string>();
-
-                foreach (var projectReference in framework.ProjectReferences)
+                // Validate all P2P references
+                foreach (var framework in project.RestoreMetadata.TargetFrameworks)
                 {
-                    if (visitedP2PReference.Add(projectReference.ProjectUniqueName))
+                    bool projectTfmFound = project.TargetFrameworks.Any(t => t.FrameworkName.Equals(framework.FrameworkName));
```

Comment:
I would use the transitive property of equality to avoid checking ProjectRestoreFrameworkInfo against TargetFrameworkInfo, and instead only do ProjectRestoreFrameworkInfo to NuGetLockFile target framework. 

At this point, you know that TargetFrameowrkInfo is equivalent to NuGetLockFile target framework.

This *kind of* feels out of place right now.